Step4 - 요금 조회 리뷰 요청드립니다.

programmer-sjk opened the pull request:
안녕하세요 동휘님.
이번 미션은 좀 난이도가 있었던 것 같은 느낌입니다.

처음에 요구사항을 이해하고 과제를 딱 시작하려 했는데 뭐 부터 해야 할지 잘 모르겠더라구요.
그래서 우선은 단계를 나눠서 `노선별 추가요금` -> `거리에 따른 추가요금` -> `나이에 따른 할인`으로 단계를 분리해서 진행하니 처음에 뭐 부터 해야하지? 에서 진행할 수 있었습니다.

처음 부딫힌 난관은 `최단경로가 여러 노선을 거칠 때 최대 추가요금`을 구하는 부분이였습니다.
최단경로 조회 결과는 `지하철 역`과 `거리`가 나오는데 이에 해당하는 노선을 구한다...?
머리속에서 아주 추잡한 로직들만 생각이 나는데 이건 아닌 것 같다고 생각했습니다.

한참을 고민해도 내 머리에서 떠오르는 로직을 구현하는 건 아닌 것 같은데.. 하다가 결국 이전분들의 로직을 참고했습니다.
대부분 `SectionEdge`를 활용했는데... 구현 방법을 몰라 참고한 건 처음인 것 같네요.. :cry: :disappointed:

기능을 다 구현한 후에 신경 쓴 부분은 `요구사항이 늘어나는 case` 였습니다.
나이에 따라 할인 정책이 더 늘어나면 수정사항이 어떻게 되지? 거리에 대한 추가요금 정책이 더 늘어나면 어떻게 되지?

`나이에 따른 할인 정책`은 `Enum의 요소 추가`로 대응할 수 있었는데 `거리에 대한 추가`는 메소드 내부에 분기를 더 추가해야 했습니다. **왜 다를까?** 고민해보니 `변경되는 부분이 Enum 안에 있냐 없냐`로 생각했습니다.

`나이에 따른 할인 정책`은 **변하는 부분**인 `할인 percent`가 Enum 마다 달리 유지되지만, `거리에 따른 요금정책`은 해당 부분이 밖에 있었거든요. 그래서 거리에 따른 요금을 계산하는 로직을 함수로 Enum 내부로 넣었습니다.

저는 실무에서 NodeJS에 함수형이 더 익숙해서, OOP에 대한 지식이 많이 부족합니다.
나이에 따른 할인 정책을 인터페이스로 할까? 라는 생각이 들었지만 굳이..? 지금도 Enum에 요소만 추가하면 유연하게 대응 할 수 있는 것 같은데 생각을 했는데, 혹시 여기에 인사이트나 의견 있으면 남겨주세요.

아 그리고 마지막 미션이네요. 단순히 제 실수나 누락 등이 아니라 리뷰를 통해 몇 가지 배울수 있어서 좋았습니다.
그럼 4단계 리뷰 부탁드릴게요 :pray:

programmer-sjk committed:
docs(guide): README.md 요구사항 정리

programmer-sjk committed:
feat(domain): 추가 요금(ExtraCharge) 도메인 추가

programmer-sjk committed:
refactor: Line 필드와 생성자에 추가요금 컬럼이 추가되면서 기존 코드 리팩토링

programmer-sjk committed:
refactor: LineRequest Builder 사용하는 곳에 추가요금 적용

programmer-sjk committed:
refactor: ExtraCharge -> ExtraFare 이름 변경

programmer-sjk committed:
test(domain): ExtraFare domain 테스트 추가

programmer-sjk committed:
refactor(test): 최단 경로 조회 시, 지하철 이용요금 검증하도록 수정

programmer-sjk committed:
feat(domain): 최단거리 그래프 생성시 SectionEdge 추가

programmer-sjk committed:
test(domain): Lines 최대 추가요금 구하는 테스트 추가

programmer-sjk committed:
feat(domain): Lines 최대 추가요금 구하는 기능 추가

programmer-sjk committed:
test(domain): Fare(요금) domain 테스트 코드 추가

programmer-sjk committed:
feat(domain): Fare(요금) domain 기능 추가

programmer-sjk committed:
refactor(test): 불필요한 변수 제거

programmer-sjk committed:
test(domain): FareDistance(거리별 요금) 테스트 추가

programmer-sjk committed:
feat(domain): FareDistance(거리별 요금) 도메인 추가 및 변경사항 적용

programmer-sjk committed:
test(domain): DiscountAge(연령별 할인) domain 테스트 코드 추가

programmer-sjk committed:
feat(domain): DiscountAge(연령별 할인) domain 기능 추가

programmer-sjk committed:
refactor: 최단 경로 조회시 나이에 따른 할인을 위해 인증 어노테이션 추가

programmer-sjk committed:
refactor: 비로그인 유저도 최단경로 검색할 수 있도록 LoginMember에 Guest 추가

programmer-sjk committed:
test(domain): 할인정책과 요금 domain에 대한 테스트 보강

programmer-sjk committed:
refactor: Fare에 Lines 대신 추가요금을 전달하는 형태로 수정

programmer-sjk committed:
refactor: 연령별 요금 할인 변수명 수정

programmer-sjk committed:
refactor: ShortestPath 에서 StationResponse 리스트 대신 Station 리스트를 리턴하도록 구조 수정

programmer-sjk committed:
refactor: PathFinder에서 GraphPath 대신 필요한 값을 전달하도록 수정

programmer-sjk committed:
refactor: ExtraFare -> Fare로 변경

programmer-sjk committed:
test(domain): ShortestPath의 요금계산 기능 테스트 추가

programmer-sjk committed:
refactor: 메소드 이름 가독성 좋게 수정

programmer-sjk committed:
refactor: FareDistance 계산값을 Enum 요소로 추가해서 변경에 영향이 적도록 리팩토링

programmer-sjk committed:
refactor: FareDistance 거리에 해당하는 Enum 객체가 없으면 NOT_MATCH를 리턴하도록 수정

programmer-sjk committed:
test(acceptance): 요금에 대한 인수테스트 구체적으로 추가 및 리팩토링

programmer-sjk committed:
refactor: 안쓰는 import 제거

programmer-sjk committed:
refactor: 문자열 에러 string -> enum 으로 관리

dhmin5693 commented on src/main/java/nextstep/subway/auth/domain/LoginMember.java:
```diff
@@ -7,6 +7,12 @@ public class LoginMember {
+        LoginMember guest = new LoginMember();
+        guest.age = 0;
+        return guest;
+    }
```
매번 새 guest 가 만들어질 필요는 없습니다.

programmer-sjk replied:
감사합니다.

dhmin5693 commented on src/main/java/nextstep/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -26,6 +27,10 @@ public boolean supportsParameter(MethodParameter parameter) {
     public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest, WebDataBinderFactory binderFactory) {
         String credentials = AuthorizationExtractor.extract(webRequest.getNativeRequest(HttpServletRequest.class));
+        if (credentials == null) {
+            return LoginMember.guest();
```
이렇게 해버리면 모든 회원용 서비스가 비회원 접근을 허용하게 됩니다.
힌트를 드리자면 `AuthenticationPrincipal` 에 로그인이 필수인지를 명시하는 required 같은 필드를 두거나 `GuestOrMember` 처럼 회원/비회원 모두가 함께 사용할 수 있는 별도의 resolver를 만드는 방법이 있습니다.

dhmin5693 commented on src/main/java/nextstep/subway/path/domain/ShortestPath.java:
```diff
@@ -0,0 +1,32 @@
+    public ShortestPath(List<Station> stations, int distance, Lines lines) {
+        this.stations = stations;
+        this.distance = distance;
+        this.lineExtraFare = lines.maxExtraFare().get();
```
https://tecoble.techcourse.co.kr/post/2020-06-02-law-of-demeter/
스트림처럼 체이닝 방식의 연산을 제외하고 한 줄에 점은 하나만 찍는게 좋습니다.

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Fare.java:
```diff
@@ -0,0 +1,32 @@
+        }
+    }
+
+    public int get() {
+        return value;
+    }
```
일반적으로 get은 데이터를 꺼내오는 행위를 의미하는 경향이 있는 것 같습니다.
사람에 따라 의견의 차이가 조금 있겠으나 개인적으로는 get 보다 value가 더 적절한 메소드명이 아닐까 싶네요.

programmer-sjk replied:
오... 네 수정해보니 value가 훨씬 좋은 것 같습니다

dhmin5693 commented on src/main/java/nextstep/subway/path/domain/AgeDiscount.java:
```diff
@@ -0,0 +1,37 @@
+
+    private final int start;
+    private final int end;
+    private final double percent;
```
percent 보다는 비율을 나타내는 rate가 조금 더 적절하지 않을까요?

dhmin5693 commented on src/main/java/nextstep/subway/path/domain/FareDistance.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Arrays;
+import java.util.function.IntFunction;
+
+public enum FareDistance {
```
enum에 너무 복잡한 로직을 넣는건 개인적으로 선호하지는 않습니다.

그 점을 빼고 생각해봐도 인터페이스 방식이 여러모로 유리할 것 같아요.
대부분의 디자인 패턴은 유연한 설계로부터 나오는데, 그 시작점이 인터페이스이기도 하구요.

큰 고민없이 떠오른 건 데코레이터 패턴 정도가 있네요.
최대 거리 시점에서 사용하는 구현체를 생성하고, 그 구현체에선 조금 더 낮은 거리의 구현체를 생성하면서 값을 전달하고 하는 방식으로 fare에 살을 붙여간다고 생각해주시면 됩니다.
물론 원하시는 방법을 쓰셔도 상관없습니다.

그 외에도 `MIDDLE`이나 `LONG`이 다른 enum의 로직을 의존하고 있는데, enum은 아무래도 상수의 개념이다 보니 의존성이 있는 이런 관계는 조금 완화를 해주는게 좋겠다는 생각도 있습니다.

dhmin5693 commented on src/main/java/nextstep/subway/path/dto/PathResponse.java:
```diff
@@ -1,16 +1,23 @@
+                .map(StationResponse::from)
+                .collect(Collectors.toList());
         this.distance = distance;
+        this.fare = fare.get();
```
어짜피 int값만 전달할 목적이라면 `Fare` 보다는 fare의 value를 직접적으로 전달해주는 편이 좋지 않을까요?
객체 사이의 연결을 최대한 줄여야 수정에 영향을 덜 받게 됩니다.

programmer-sjk replied:
사실 int 값을 전달할지 Fare를 전달할지 조금 고민했던 부분이였는데
수정을 생각해보면 객체를 전달하지 않는게 더 좋네요.

dhmin5693 reviewed (changes requested):
안녕하세요.

이번 미션 요구사항이 전반적으로 어려운 편입니다.
다른 사람의 구현을 보고 참고하는 것도 좋은 공부 방법이에요.

궁금해하시는 부분에 답을 남겼습니다.
그리고 수정이 필요한 부분에도 의견을 몇 개 남겨보았으니 확인해보세요.
수정 완료되면 다시 리뷰 요청 부탁드리겠습니다.

programmer-sjk committed:
refactor: 코드 리뷰 반영

코드 리뷰 반영
- 변수명 수정
- 디미터 법칙 반영
- Fare add, subtract 추가

programmer-sjk committed:
refactor: 코드 리뷰 반영 (최단경로 비로그인)

코드 리뷰 반영
- AuthenticationPrincipal에 required 필드 추가
- 최단경로 controller에 required = false 추가
- Resolver 내부에서 required가 false, 인증값이 없다면 Guest 동작하도록 수정
- LoginMember에서 Guest 캐시

programmer-sjk committed:
refactor: 코드 리뷰 반영 (데코레이터 패턴 적용)

programmer-sjk commented:
동휘님 안녕하세요.

저는 리뷰어 분들이 제가 모르는 방식을 제안하면 항상 그렇게 하는 편입니다.
모르는 방식 자체가 제가 해보면서 배울수 있다고 생각하기 때문인데요.

데코레이터 패턴을 검색해보니 아주 오래전에 Head First 디자인 패턴 책에서 봤던 것 같은 패턴이네요.
처음 구현해보는데, 의도하신 대로 제가 구현했을지 조금 걱정되네요.

확인 부탁드립니다. 🙏

dhmin5693 commented:
리뷰에도 써둔건데 얼핏 생각나는 방법을 말했을 뿐입니다.
그래서 더 좋은 방법 있으면 그걸 쓰시라고 안내를 드렸어요.
구현은 되게 잘 해주셨네요.

앞으로 리뷰도 마찬가지인 내용이지만 모르는 내용을 안내받았다 한들 무작정 받아들이시기보다는
내 상황에 진짜 괜찮은 내용일까?
그렇지 않다면 무엇이 문제인가?
소통이 잘 안 되었나?
나에게 혹은 지금 상황에 적용이 어려운가?
리뷰어가 잘못된 의견을 준걸까?

이런걸 한번씩 생각해보시고 리뷰어와 토론을 해보시는게 더 좋습니다.
리뷰어와 수강생 모두 경력이 있는 개발자니까요.

dhmin5693 reviewed (approved):
피드백 반영 확인했습니다. :)
어려운 요구사항이 많았지만 잘 헤쳐나가주셨네요.

어려운 미션 수행하느라 정말 고생 많으셨습니다.
여기까지 오셨다면 앞으로 코드 미션은 하나 남아서 다소 편해지겠군요.

남은 미션도 잘 마무리하시길 바랍니다.
정국님의 완주를 응원하겠습니다.
